kernel/src/hil/hw_debug.rs

```diff
@@ -0,0 +1,31 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Interfaces for interacting with debug hardware integrated in various SoCs.
+//! Currently allows reading the cycle counter.
+
+pub trait CycleCounter {
+    /// Enable and start the cycle counter.
+    fn start(&self);
```

Comment:
Should this return a `Result`? It seems legal to not implement the counters in RISC-V

Reply:
What are the errors? If there is no counter then we can't implement this trait.

Reply:
This is a bit tricky though. Would a chip implement the trait? That way we know at build time if it is supported or not. But if the architecture implements the trait (like ARM does) we don't know if it's supported or not

Reply:
Wherever the code is, it can only be instantiated if the hardware exists.

Reply:
Agree with brad -- even if the implementations generally live in the arch/ crate, only boards that actually support that feature of the arch should instantiate something that implements this trait.

Reply:
Yeah I realise that's the right way.

The only other issue is what happens if different revisions of a board do/don't implement the feature. But hopefully that never happens

Reply:
I assume different revisions of a board would still require a different top-level `main.rs`, or some way to modify what is included in that main.rs at compile time